### PR TITLE
sample_channel: Create pin names from configuration string and add pinname to sample

### PR DESCRIPTION
**Pull request by luminize**
Since instantiatiable components can already have a chosen name, this adds the ability to create pins with configurable names, allowing the component instance to mimic a component. (Only IN pins for now, OUT pins reading values from a ring will come later)

You can mimick behaviour of a driver, or testing a HAL config without the hardware, like the linuxcnc ethercat driver, then can be done in python.

The sample now also holds the pin name, this is for necessary for keeping track of values of different pins, for example when you save data in a CSV file

### CODE REVIEW

File: src/hal/sample_channel/sample_channel.hal [hunk 1]
```diff
@@ -9,13 +9,17 @@ loadrt siggen
 # tell max cycles the ring should be able to contain by: cycles=100
 newinst sample_channel_pb sampler --- samples=bfusUSf cycles=100
 
+newinst sample_channel_pb lcec.0.1 --- samples=bff pinnames=food,bar,bazooka cycles=100
```
Comment (ArcEye) [inline src/hal/sample_channel/sample_channel.hal]: The third `-` is probably being passed as an arg which is why it mostly works.
Should only need 2 `--` for hal_command to separate off the strings after them and pass as args to the component.

Comment (luminize) [reply src/hal/sample_channel/sample_channel.hal]: hmm, I've taken the debug output from both instaces, I had seen this but not made the link that this was not meant to be. I've always used 3 dashes, probably slipped in a typo after one of the first examples I've seen when it got introduced.
```
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt halg_xinitfv:90 HAL: initializing component 'sample_channel_pb' type=1 arg1=0 arg2=0/0x0
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt sample_channel_pb: sampler: argument 2 = ---
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt sample_channel_pb: sampler: argument 3 = samples=bfusUSf
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt sample_channel_pb: sampler: argument 4 = cycles=100
...
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt sample_channel_pb: lcec.0.1: argument 2 = ---
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt sample_channel_pb: lcec.0.1: argument 3 = samples=bff
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt sample_channel_pb: lcec.0.1: argument 4 = pinnames=food,bar,bazooka
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt pinnames = food,bar,bazooka
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt sample_channel_pb: lcec.0.1: argument 5 = cycles=100
Jun  9 13:44:56 TP51 msgd:0: hal_lib:1807:rt cycles = 100
```
and indeed, changing the `---` to `--` shows:
```
Jun  9 13:55:11 TP51 msgd:0: hal_lib:3352:rt halg_xinitfv:90 HAL: initializing component 'sample_channel_pb' type=1 arg1=0 arg2=0/0x0
Jun  9 13:55:11 TP51 msgd:0: hal_lib:3352:rt sample_channel_pb: sampler: argument 2 = samples=bfusUSf
Jun  9 13:55:11 TP51 msgd:0: hal_lib:3352:rt sample_channel_pb: sampler: argument 3 = cycles=100
...
Jun  9 13:55:11 TP51 msgd:0: hal_lib:3352:rt sample_channel_pb: lcec.0.1: argument 2 = samples=bff
Jun  9 13:55:11 TP51 msgd:0: hal_lib:3352:rt sample_channel_pb: lcec.0.1: argument 3 = pinnames=food,bar,bazooka
Jun  9 13:55:11 TP51 msgd:0: hal_lib:3352:rt pinnames = food,bar,bazooka
Jun  9 13:55:11 TP51 msgd:0: hal_lib:3352:rt sample_channel_pb: lcec.0.1: argument 4 = cycles=100
```
i'll fix the hal file

**Comment by luminize** [comment]
I've been looking through the code to fix these, but I see some references that 3 dashes are to be used:
https://github.com/machinekit/machinekit-hal/blob/cfa5d850b463dd01dcdf5ef78e8d4c8075a7e149/src/hal/utils/instcomp.g#L430-L434
https://github.com/machinekit/machinekit-hal/blob/77089bab72d670e5b8060f50e69049ceb13145f6/src/hal/user_icomps/watch.c#L18

So, are these mentions wrong? if so, I'll make a separate PR to fix that.

**Comment by ArcEye** [comment]
> So, are these mentions wrong?

Damn, yes they are :(

The problem only arises if you make assumptions about argument numbers.
If you just iterate through args until you find the one you want, it does not show up, because you skip any meaningless dash(es).

If you are OK fixing those that's great, otherwise I will clear up my own mess!

**Comment by luminize** [comment]
> On 9 Jun 2019, at 17:24, ArcEye <notifications@github.com> wrote:
> 
> If you are OK fixing those that's great, otherwise I will clear up my own mess!
> 

No problem, I’ll fix them in another pr

**Pull request merged**